### PR TITLE
feat(common): add segmenter for element strings 🙀

### PR DESCRIPTION
**Pull request by srl295**
For #7377

- detect and reject nested square brackets, for now
- implement segmentation of reorder element strings,
  including type identification (escape, codepoint, uset)
- tests

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by srl295** [comment]
this would allow fallback to a more complex (recursive) analysis of nested square brackets in the future

**Comment by srl295** [comment]
trying to keep the non-kmxplus-common part here.

File: common/web/types/test/util/test-unescape.ts [hunk 1]
```diff
@@ -56,3 +56,9 @@ describe('test unescapeString()', function() {
     assert.throws(() => unescapeString('\\u{110000}'), UnescapeError);
   });
 });
+
+describe('test unescapeOneQuadString()', () => {
+  it('should be able to convert', () => {
+    assert.equal(unescapeOneQuadString('\\u0127'), '\u{0127}');
```
Comment (jahorton) [inline common/web/types/test/util/test-unescape.ts:62]: Might be a little clearer on the surface if the second string were simply `'ħ'`.

I was fairly confused for a bit until I realized that the critical detail here is that the former case has the escaping-character escaped, but not the latter.

File: common/web/types/src/util/util.ts [hunk 2]
```diff
@@ -19,12 +19,47 @@ export function boxXmlArray(o: any, x: string): void {
 // TODO-LDML: #7569 the below regex works, but captures more than it should
 // (it would include \u{fffffffffffffffff } which
 // is overlong and has a space at the end.) The second regex does not work yet.
-const MATCH_HEX_ESCAPE = /\\u{([0-9a-fA-F ]{1,})}/g;
+export const MATCH_HEX_ESCAPE = /\\u{([0-9a-fA-F ]{1,})}/g;
 // const MATCH_HEX_ESCAPE = /\\u{((?:(?:[0-9a-fA-F]{1,5})|(?:10[0-9a-fA-F]{4})(?: (?!}))?)+)}/g;
 
+/** regex for single quad escape such as \u0127 */
+export const MATCH_QUAD_ESCAPE = /\\u([0-9a-fA-F]{4})/g;
```
Comment (jahorton) [inline common/web/types/src/util/util.ts:26]: Will this match the alternate escaping pattern referred to in the user test?  it doesn't seem like this regex would match `'\u{0127}'`, but it's a perfectly valid JS quad-escape, isn't it?

File: common/web/types/test/util/test-unescape.ts (hunk 1 above)
Comment (mcdurdin) [reply common/web/types/test/util/test-unescape.ts:62]: Alternative, just add a comment:

```suggestion
    // testing that `\u0127` is unescaped correctly (to U+0127: 'ħ')
    assert.equal(unescapeOneQuadString('\\u0127'), '\u{0127}');
```

**Review by mcdurdin — APPROVED** [review]

File: common/web/types/src/util/util.ts (hunk 2 above)
Comment (srl295) [reply common/web/types/src/util/util.ts:26]: No that is handled by the previous escape. This is for a special one that we might remove with no curly braces

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.133-alpha](https://keyman.com/downloads/releases/alpha/17.0.133)

**Pull request merged**